core/src/main/scala/scalaz/Frap.scala

```diff
@@ -0,0 +1,118 @@
+package scalaz
+
+/**
+ * Free applicative functors. Less expressive than free monads, but more
+ * flexible to inspect and interpret.
+ */
+sealed trait Frap[F[_],A] {
+  import Frap._
+
+  /**
+   * The canonical natural transformation that interprets this free
+   * program by giving it the semantics of the applicative functor `G`.
+   * Not tail-recursive unless `G` is a free monad.
+   */
+  def foldMap[G[_]:Applicative](f: F ~> G): G[A] =
+    this match {
+      case Pure(x) => Applicative[G].pure(x)
+      case x@Ap() => Applicative[G].ap(f(x.v()))(x.k() foldMap f)
+    }
+
+  /** Provides access to the first instruction of this program, if present */
+  def para[B](pure: A => B, ap: ({type λ[α] = (F[α], Frap[F, α => A])})#λ ~>
```

Comment:
I can't find this on GH, replying by email. Yes, KP can do that via the
type function syntax, eg: `λ[α => (α, α)]`

On 8 January 2015 at 01:00, Rúnar <notifications@github.com> wrote:

> In core/src/main/scala/scalaz/Frap.scala
> <https://github.com/scalaz/scalaz/pull/872#discussion-diff-22586681>:
>
> > +sealed trait Frap[F[_],A] {
> > +  import Frap._
> > +
> > +  /**
> > +   * The canonical natural transformation that interprets this free
> > +   * program by giving it the semantics of the applicative functor `G`.
> > +   * Not tail-recursive unless `G` is a free monad.
> > +   */
> > +  def foldMap[G[_]:Applicative](f: F ~> G): G[A] =
> > +    this match {
> > +      case Pure(x) => Applicative[G].pure(x)
> > +      case x@Ap() => Applicative[G].ap(f(x.v()))(x.k() foldMap f)
> > +    }
> > +
> > +  /** Provides access to the first instruction of this program, if present */
> > +  def para[B](pure: A => B, ap: ({type λ[α] = (F[α], Frap[F, α => A])})#λ ~>
>
> Wait, what? Oh, can KP not support f[x] = (x,x) ?
> … <#14ac4b24149d2cc1_>
> On Tue Jan 06 2015 at 11:02:09 PM Jed Wesley-Smith <
> notifications@github.com> wrote: In core/src/main/scala/scalaz/Frap.scala
> <https://github.com/scalaz/scalaz/pull/872#discussion-diff-22570127>: >
> +sealed trait Frap[F[_],A] { > + import Frap._ > + > + /** > + * The
> canonical natural transformation that interprets this free > + * program by
> giving it the semantics of the applicative functor `G`. > + * Not
> tail-recursive unless `G` is a free monad. > + */ > + def
> foldMap[G[_]:Applicative](f: F ~> G): G[A] = > + this match { > + case
> Pure(x) => Applicative[G].pure(x) > + case x@Ap() =>
> Applicative[G].ap(f(x.v()))(x.k() foldMap f) > + } > + > + /** Provides
> access to the first instruction of this program, if present */ > + def
> para[B](pure: A => B, ap: ({type λ[α] = (F[α], Frap[F, α => A])})#λ ~>
> bastard! #875 <https://github.com/scalaz/scalaz/pull/875> <
> https://github.com/scalaz/scalaz/pull/875> — Reply to this email directly
> or view it on GitHub <
> https://github.com/scalaz/scalaz/pull/872/files#r22570127>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scalaz/scalaz/pull/872/files#r22586681>.
>